marlowe-chain-sync/marlowe-protocols/src/Network/Protocol/FilteredChainSync/Types.hs

```diff
@@ -0,0 +1,134 @@
+
+module Network.Protocol.FilteredChainSync.Types where
+
+import Data.Binary (Binary (..))
+import Data.Kind (Type)
+import Data.String (IsString)
+import Data.Text (Text)
+import qualified Data.Text.Encoding as T
+import Network.TypedProtocol (Protocol (..))
+
+-- | The type of states in the protocol.
+data FilteredChainSync (query :: Type -> Type -> Type) point tip where
+  -- | The server is waiting for the client to initiate the handshake.
+  StInit :: FilteredChainSync query point tip
+
+  -- | The client is waiting for the server to accept the handshake.
+  StHandshake :: FilteredChainSync query point tip
+
+  -- | The client and server are idle. The client can send a request.
+  StIdle :: FilteredChainSync query point tip
+
+  -- | The client has sent a next update request. The client is now waiting for
+  -- a response, and the server is preparing to send a response. The server can
+  -- respond immediately or it can send a 'Wait' message followed by a response
+  -- at some point in the future.
+  StNext :: err -> result -> StNextKind -> FilteredChainSync query point tip
+
+  -- | The failed state of the protocol.
+  StFault :: FilteredChainSync query point tip
+
+  -- | The normal terminal state of the protocol.
+  StDone :: FilteredChainSync query point tip
+
+-- | Sub-states of 'StNext'.
+data StNextKind
+  = StCanAwait  -- ^ The server can send a 'Wait' message or a response
+  | StMustReply -- ^ The server must send a reply, having sent a 'Wait'.
+
+-- | Schema version used for
+newtype SchemaVersion = SchemaVersion Text
+  deriving stock (Show, Eq, Ord)
+  deriving newtype (IsString)
+
+instance Binary SchemaVersion where
+  put (SchemaVersion v) = put $ T.encodeUtf8 v
+  get = do
+    bytes <- get
+    case T.decodeUtf8' bytes of
+      Left err      -> fail $ show err
+      Right version -> pure $ SchemaVersion version
+
+instance Protocol (FilteredChainSync query point tip) where
+
+  -- | The type of messages in the protocol. Corresponds to the state
+  -- transition in the state machine diagram.
+  data Message (FilteredChainSync query point tip) from to where
+
+    -- | Initiate a handshake for the given schema version.
+    MsgRequestHandshake :: SchemaVersion -> Message (FilteredChainSync query point tip)
+      'StInit
+      'StHandshake
+
+    -- | Accept the handshake.
+    MsgConfirmHandshake :: Message (FilteredChainSync query point tip)
+      'StHandshake
+      'StIdle
+
+    -- | Reject the handshake.
+    MsgRejectHandshake :: [SchemaVersion] -> Message (FilteredChainSync query point tip)
+      'StHandshake
+      'StFault
+
+    -- | Request the next matching result for the given query from the client's
+    -- position.
+    MsgQueryNext :: query err result -> Message (FilteredChainSync query point tip)
+      'StIdle
+      ('StNext err result 'StCanAwait)
+
+    -- | Reject a query with an error message.
+    MsgRejectQuery :: err -> tip -> Message (FilteredChainSync query point tip)
```

Comment:
- [ ] TODO remove and allow specific queries to represent failure with an `Either`